### ISSUE TITLE
touch creating files newer than system clock

### ISSUE THREAD

**Issue by Finermeerkat137** [uutils/coreutils#1592]
I tried using the uutils and was compiling another piece of software from source with the `./configure && make && make install` dance, when I got an error saying that the files were newer than system clock. Removing uutils from the PATH seemed to fix the issue. Is it a problem with `touch`?

**Comment by sylvestre** [comment]
Would it be possible for you to provide some step to reproduce?
It is too vague to be actionable now

**Comment by chadbrewbaker** [comment]
Could you run this and paste the results?

```bash
strace uutils_touch /tmp/newfile
```

**Comment by Finermeerkat137** [comment]
Yeah, sure :) Just to clarify, here's what I get when I try to compile `sed` from the tarball:
```
checking for a BSD-compatible install... /opt/uutils/bin/install -c
checking whether build environment is sane... configure: error: ls -t appears to fail.  Make sure there is not a broken
  alias in your environment
configure: error: newly created file is older than distributed files!
Check your system clock
```

**Comment by Finermeerkat137** [comment]
Heres gnu t

**Comment by chadbrewbaker** [comment]
GNU touch

```bash
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=3041456, ...}) = 0
mmap(NULL, 3041456, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f6ca71d9000
close(3)                                = 0
openat(AT_FDCWD, "/tmp/newfile", O_WRONLY|O_CREAT|O_NOCTTY|O_NONBLOCK, 0666) = 3
dup2(3, 0)                              = 0
close(3)                                = 0
utimensat(0, NULL, NULL, 0)             = 0
close(0)                                = 0
close(1)                                = 0
close(2)                                = 0
```
uutils touch
```bash
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3576, ...}) = 0
statx(0, NULL, AT_STATX_SYNC_AS_STAT, STATX_ALL, NULL) = -1 EFAULT (Bad address)
statx(AT_FDCWD, "/tmp/newfile", AT_STATX_SYNC_AS_STAT, STATX_ALL, 0x7ffd9867ea20) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/newfile", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 3
fcntl(3, F_GETFD)                       = 0x1 (flags FD_CLOEXEC)
close(3)
```
Why is is calling a stat on /etc/localtime?

**Comment by chadbrewbaker** [comment]
Problems I see:
1) uutils_touch is truncating files by default - that's a problem?
2) uutils_touch should also call utimensat()?

I'll run with code coverage to make sure I understand everything it is calling.

**Comment by rivy** [comment]
So, I think that, in order to proceed, we need to nail down whether this is a `touch` issue or rust library issue.

My quick look at the `touch` code reveals nothing that looks incorrect about its logic or implementation. The integration tests are testing (and passing) for time correctness when setting the file times. And I see no obvious correlation in the rust code to the lower-level syscall differences. A more specific failure at the command line level would be helpful to design a test and move forward. Is there a way to more exactly tease out the process and failure condition of the failing compilation?

**Comment by chadbrewbaker** [comment]
Library issue that hard codes the truncate.

https://github.com/rust-lang/rust/blob/4ffb5c5954a304daf47a567b34e74e421db86d98/library/std/src/fs.rs#L357

```rust
#[stable(feature = "rust1", since = "1.0.0")]
    pub fn create<P: AsRef<Path>>(path: P) -> io::Result<File> {
        OpenOptions::new().write(true).create(true).truncate(true).open(path.as_ref())
    }
```

GNU discussion on their design decisions for what flags to set: https://lists.gnu.org/archive/html/bug-coreutils/2018-10/msg00340.html

**Comment by rivy** [comment]
I read through the referenced thread, but I don't see any direct discussion of the truncate option, unless it somehow equates with closing the file descriptor?

So, is the "truncate()" actually a problem? If it is, we could open the file using the same lower level code minus the truncate call (eg, `OpenOptions::new().write(true).create(true).open(path.as_ref())`). But, is it the cause of this issue?

**Comment by kevinburkemeter** [comment]
Just following up - is it still true that `touch` truncates files instead of just updating their mtime?

**Comment by Arcterus** [comment]
@kevinburkemeter the line that truncates files only does so if `touch` believes the file doesn't actually exist.  There is a race condition in that the file could be created between the `if` condition and the call to `File::create()`, but generally you're not going to have a problem if you're using the utility normally.

**Comment by ChrisDryden** [comment]
I may not be understanding the full issue here, but to me it seems that this was able to fix the issue? https://github.com/uutils/coreutils/pull/9870
